pkg/operator/apiserver/controller/workload/workload.go

```diff
@@ -287,9 +287,26 @@ func (c *Controller) updateOperatorStatus(ctx context.Context, previousStatus *o
 		desiredReplicas = *(workload.Spec.Replicas)
 	}
 
+	selector, err := metav1.LabelSelectorAsSelector(workload.Spec.Selector)
+	if err != nil {
+		return fmt.Errorf("failed to construct label selector: %v", err)
+	}
+	matchingPods, err := c.podsLister.Pods(c.targetNamespace).List(selector)
+	if err != nil {
+		return err
+	}
+	// Terminating pods don't account for any of the other status fields but
+	// still can exist in a state when they are accepting connections and would
+	// contribute to unexpected behavior if we report Progressing=False.
+	// The case of too many pods might occur for example if `TerminationGracePeriodSeconds` is set.
+	terminatingPods := countTerminatingPods(matchingPods)
+
 	// If the workload is up to date, then we are no longer progressing
 	workloadAtHighestGeneration := workload.ObjectMeta.Generation == workload.Status.ObservedGeneration
-	workloadIsBeingUpdated := workload.Status.UpdatedReplicas < desiredReplicas
+	// update is done when all pods have been updated to the latest revision
+	// and the deployment controller has reported NewReplicaSetAvailable
+	workloadIsBeingUpdated := !workloadAtHighestGeneration || !hasDeploymentProgressed(workload.Status)
+	workloadHasOldTerminatingPods := workload.Status.AvailableReplicas == desiredReplicas && terminatingPods > 0
```

Comment:
In theory this math could indicate a terminating pod that's not actually "old", right?

Like, `desiredReplicas == 3` and I have `AvailableReplicas == 3`, but one of the available replicas is an old pod, and one of the terminating pods is not "old".

Reply:
It mostly works (`workload.Status.AvailableReplicas` does not include terminating replica), but yeah there is one edge case. If you delete a new pod and it terminates for a longer time then the new one takes to become available. Then this status would be incorrect.

I originally wanted to bring in the deployment controller logic in, but in the end decided against it. It would require a lot of code for matching pods/hashes and revisions/replicasets. And I would prefer not to maintain this logic.

I have noticed that all of the uses of the WorkloadController use `EnsureAtMostOnePodPerNode` function in the following repos:
- https://github.com/openshift/cluster-authentication-operator
- https://github.com/openshift/cluster-openshift-apiserver-operator

When we depend on the `DeploymentProgressing` condition from now on and `EnsureAtMostOnePodPerNode` function, we first terminate the pod on each node before we create a new one.

So I have removed the terminating pods logic and added a note that this is a user of this controller responsibility to ensure if old terminating pods can be a part of the complete deployment or not. Which is done by using `EnsureAtMostOnePodPerNode` together with this PR. Or it can be also done by a `DeploymentPodReplacementPolicy` feature once it is merged upstream.